<a href="https://colab.research.google.com/github/KenStanley/FunForPeterAndKen/blob/master/Copy_of_NBVP_model_predict.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Name Based Voting Prediction
Edit this 

AnythingInGithub 

Text explanation of what this does and how to use it.

 git remote is a .git file 

 https://github.com/KenStanley/classifyDems.git

 wht url 


Include file format expectations.

In [0]:
!pwd
%cd /content/classifyDems/
!git pull 

/content/classifyDems
/content/classifyDems
remote: Enumerating objects: 4, done.
remote: Counting objects: 100% (4/4), done.
remote: Compressing objects: 100% (3/3), done.
remote: Total 3 (delta 1), reused 0 (delta 0), pack-reused 0
Unpacking objects: 100% (3/3), done.
From https://github.com/KenStanley/classifyDems
   55c89c8..716504e  master     -> origin/master
Updating 55c89c8..716504e
Fast-forward
 Copy_of_NBVP_model_predict.ipynb | 3035 ++++++++++++++++++++++++++++++++++++++
 1 file changed, 3035 insertions(+)
 create mode 100644 Copy_of_NBVP_model_predict.ipynb


In [11]:
!pwd
!git add "Copy of NBVP-model_predict.ipynb"


/content/classifyDems
fatal: pathspec 'Copy of NBVP-model_predict.ipynb' did not match any files


I want to save this within github 

In [0]:
!git clone https://KenStanley:Ken4Peter@github.com/KenStanley/classifyDems.git

Cloning into 'classifyDems'...
remote: Enumerating objects: 18, done.
remote: Counting objects: 100% (18/18), done.
remote: Compressing objects: 100% (17/17), done.
remote: Total 18 (delta 4), reused 0 (delta 0), pack-reused 0
Unpacking objects: 100% (18/18), done.


## Header Section
Import additional libraries to use for data processing and prediction

In [8]:
from google.colab import drive
drive.mount('/content/drive')

Go to this URL in a browser: https://accounts.google.com/o/oauth2/auth?client_id=947318989803-6bn6qk8qdgf4n4g3pfee6491hc0brc4i.apps.googleusercontent.com&redirect_uri=urn%3aietf%3awg%3aoauth%3a2.0%3aoob&response_type=code&scope=email%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdocs.test%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdrive%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdrive.photos.readonly%20https%3a%2f%2fwww.googleapis.com%2fauth%2fpeopleapi.readonly

Enter your authorization code:
··········
Mounted at /content/drive


In [12]:
import os
print(os.getcwd())
print("Some changes")
os.chdir("./drive/My Drive/KenS_Vote_Project")

/content/classifyDems
Some changes


FileNotFoundError: ignored

In [0]:
import sys
import numpy as np
import pandas as pd
import sklearn.base
import math
import datetime
import os

print("Python version: {}.{}.{}".format(sys.version_info[0], sys.version_info[1], sys.version_info[2]))
pythonVersion = str(sys.version_info[0])+"."+str(sys.version_info[1])
print("NumPy version: ",np.__version__)
print("Pandas version: ",pd.__version__)
print("SK Learn version: ",sklearn.__version__)

# NBVP Internal variables
_MODE = {'train':True, 'predict': True} # Code will leave as 'train' or update to 'predict' depending on the input file contents
_ERROR = False
_REGCOL = 'registered'
# These Variables are available for user control.
_USE_REPO = False
_SRWOE = 'SR'#SR or WoE
_DEF_SR_F = .6 # SR values range from 0 to 1
_DEF_SR_L = .5 # F = First name, L = Last Name
_DEF_WOE_F = 1 # WOE values or + or -, typically between +/-7 
_DEF_WOE_L = 0

Python version: 3.6.9
NumPy version:  1.18.3
Pandas version:  1.0.3
SK Learn version:  0.22.2.post1


***
## Load a csv file

1. Assume labeled and unlabeled data
2. Registered or LALVOTERID columns
3. Dtypes in columns we keep.  

- Where you see `filename = ` in the next cell, change the path and filename to match the csv dataset file.
- If the new filename has different column names, that will break the code.

The specified file must include the folowing columns, specified by column headers in the first row of the csv file (though order is not important). __Text case__ is also not important.

- PARTY AFFILIATION (target) <D, R, NA, blank> Consider D and R are 2 possible labels. NA and blank are to be predicted. Drop anything else.
- AA"  - Might be useful (but not nearly as useful as it was last time )                      
- percentBlack  - the percent of the people in this census block (aveaged with surrounding blocks) who self report as black
- directPercentBlack  - the percent of the people in this census block who self report as black
- L2percentBlack - the number of people list as AA in te L2 data file in this census blocks            
- firstNameNumVoters  - the number of voters with this first name who are registered as either D or R
- firstNamefractionDem- the fraction of voters with this first name who are registered as D (as opposed to R)
- lastNameNumVoters - - the number of voters with this first name who are registered as either D or R
- lastNamefractionDem- the fraction of voters with this last name who are registered as D (as opposed to R)    
**Next rev of data file**
- age (ordinal)
- gender (categorical)
- Apartment dweller (tend Democratic )
- first name frequency - raw ( I don't expect this to be useful but it will allow you to bin it differently)
- first name frequency binned into three bins: 1 (i.e. unique) 2-4 (i.e. rare) 5+ (i.e. less rare to common)
-    In my experience, uncommon first names tend to be Democratic
- Another first name scoring
- Another ethnic grouping 


*TODO:*  
*1. Verify blank. If we keep blank to predict, we need to keep the registration column so user knows they're not registered.*  
*2. Update column info above after looking at Dtypes*

In [0]:
filename = "./NBVP_data_in/voterMatchWithMoreData.csv" # All columns, labeled and unlabeled # Path for working locally

voterInfoDF = pd.read_csv(filename, index_col = 0)

#Note: There may be a warning here about mixed data types in some of the columns. That's ok as long as those columns 
#aren't the ones we're interested in. They'll get dropped and we'll check the datatypes on ours.

In [0]:
voterInfoDF.head()
voterInfoDF.columns

Index(['Name.First', 'fullAddress', 'Name.Last', 'censusId', 'SOS_VOTERID',
       'Individual.Id', 'Census.Tract', 'Census.Block', 'Census.Block.Group',
       'Latitude', 'Longitude', 'Ethnic.Group', 'Gender', 'Inferred.Age',
       'Unit.Designator', 'Unit.Designator.Number', 'Primary.Address',
       'Secondary.Address', 'Secondary.Address.Present', 'totalScore',
       'nameAndBirthScore', 'registeredToVote', 'PARTY_AFFILIATION',
       'LAST_NAME', 'FIRST_NAME', 'RESIDENTIAL_ADDRESS1',
       'RESIDENTIAL_SECONDARY_ADDR', 'PRECINCT_NAME', 'PRECINCT_CODE', 'count',
       'blockAA', 'percentBlack', 'directPercentBlack', 'L2percentBlack',
       'firstNameOccurence', 'firstNameNumVoters', 'firstNamefractionDem',
       'firstNameOccurenceBinned', 'lastNameNumVoters', 'lastNamefractionDem',
       'apartmentDwellerPerVF', 'apartmentDwellerPerL2', 'apartmentDweller',
       'inVoterFile', 'AA', 'Hispanic', 'Asian', 'Arab', 'Jewish',
       'Mediterranean', 'democrat', 'demsInHouse', 

**To change the list of columns kept, update the <code>colsToKeep</code> list here**

In [0]:
currentCols = list(voterInfoDF.columns)
colsToKeep = ['SOS_VOTERID',
       'Individual.Id','PARTY_AFFILIATION', 'AA', 'percentBlack', 'firstNameNumVoters','firstNamefractionDem', 
              'lastNameNumVoters', 'lastNamefractionDem', 'Inferred.Age', 'Gender', 'apartmentDweller',
              'firstNameOccurenceBinned', 'Hispanic', 'Asian', 'Arab', 'Jewish', 'Mediterranean']
# Considering: directPercentBlack, L2percentBlack, firstNameOccurence
colsToDrop = [col for col in currentCols if col not in colsToKeep]
# The statement above is the pythonic, efficient way of doing the following:
#colsToDrop = []
#for col in currentCols:
#    if col not in colsToKeep:
#        colsToDrop.append(col)
voterInfoDF.drop(columns = colsToDrop, inplace = True)

In [0]:
voterInfoDF.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 20493 entries, 1 to 20493
Data columns (total 18 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   SOS_VOTERID               12586 non-null  object 
 1   Individual.Id             20493 non-null  int64  
 2   Gender                    20064 non-null  object 
 3   Inferred.Age              18611 non-null  float64
 4   PARTY_AFFILIATION         4643 non-null   object 
 5   percentBlack              20489 non-null  float64
 6   firstNameNumVoters        20233 non-null  float64
 7   firstNamefractionDem      20233 non-null  float64
 8   firstNameOccurenceBinned  20493 non-null  int64  
 9   lastNameNumVoters         20232 non-null  float64
 10  lastNamefractionDem       20232 non-null  float64
 11  apartmentDweller          20493 non-null  int64  
 12  AA                        20493 non-null  int64  
 13  Hispanic                  20493 non-null  int64  
 14  Asian 

**Using `.describe` we can see what's going on with our numerical types.**
* If a count value is < 20493 (the Range Index shown by `.info`, above), there are NaN (missing) values in those columns. For those, we'll have to either drop the record or impute a value using various methods.
* Some columns are not normalized/standardized. We'll do that for those columns, depending on the model.
* Mean, std, min, max values look pretty reasonable so we won't go into detail looking for outliers.

In [0]:
voterInfoDF.describe()

Individual.Id  Inferred.Age  ...        Jewish  Mediterranean
count   2.049300e+04  18611.000000  ...  20493.000000   20493.000000
mean    6.196421e+08     50.957982  ...      0.023032       0.012931
std     2.285794e+08     18.340058  ...      0.150010       0.112981
min     4.210599e+06     20.000000  ...      0.000000       0.000000
25%     5.155040e+08     35.000000  ...      0.000000       0.000000
50%     7.136664e+08     50.000000  ...      0.000000       0.000000
75%     7.842173e+08     64.000000  ...      0.000000       0.000000
max     8.496407e+08     99.000000  ...      1.000000       1.000000

[8 rows x 15 columns]

**`PARTY_AFFILIATION` is our target column.**  
Records with R and G are labeled and used for training. Records with missing or NA are to be predicted (our unlabled data). Records marked G will be dropped.

In [0]:
print("Specified values in PARTY_AFFILIATION: ")
print(voterInfoDF["PARTY_AFFILIATION"].value_counts())
print("Missing or NA values: ", voterInfoDF[voterInfoDF["PARTY_AFFILIATION"].isna()].index.size)
print("Will drop these records:")
voterInfoDF[voterInfoDF["PARTY_AFFILIATION"] == "G"]

Specified values in PARTY_AFFILIATION: 
D    2442
R    2191
G      10
Name: PARTY_AFFILIATION, dtype: int64
Missing or NA values:  15850
Will drop these records:


SOS_VOTERID  Individual.Id  Gender  ...  Arab Jewish  Mediterranean
3647   OH0021582489      515501427  Female  ...     0      0              0
4422   OH0012720733      819333729    Male  ...     0      0              0
5006   OH0012701917      104022878  Female  ...     0      1              0
9463   OH0022169244      754269888    Male  ...     0      0              0
10421  OH0012712185      515502541  Female  ...     0      0              0
11083  OH0015096531      784227724    Male  ...     0      0              0
12424  OH0018664709      732405155    Male  ...     0      0              0
14528  OH0019471529      515508123    Male  ...     0      0              0
18085  OH0012671195      634385259    Male  ...     0      0              0
20214  OH0012671925      473780507    Male  ...     0      0              0

[10 rows x 18 columns]

In [0]:
voterInfoDF[voterInfoDF["SOS_VOTERID"] == "OH0021582489"]

SOS_VOTERID  Individual.Id  Gender  ...  Arab Jewish  Mediterranean
3647  OH0021582489      515501427  Female  ...     0      0              0

[1 rows x 18 columns]

In [0]:
voterInfoDF.drop(index = voterInfoDF[voterInfoDF["PARTY_AFFILIATION"] == "G"].index, inplace = True)

In [0]:
voterInfoDF.shape

(20483, 18)

**Force all column headers and contents to lowercase** *Maybe later*

In [0]:
'''
renameDict = {}
for column in dataDF.columns:
    if dataDF[column].dtype == 'O':
        dataDF[column] = dataDF[column].str.lower()
    renameDict[column] =column.lower()
# Also need to rename columns in lowercase
dataDF.rename(columns=renameDict, inplace=True)
'''

**Drop duplicates and split into labeled and not labeled data**  
In the full file (all columns), it looks like there were 238 dupes (119 pairs). Dropping the unused columns only increases that to 143 pairs (increase of 24 pairs). In some of those 24 pairs, they look like different records when viewing all columns. In those cases, it would require somebody that knows more about how the calculated columns were derived to determine if there's a problem or which row to keep. Since we only drop 1 of each pair, that's only 12 lost records. We'll just drop them unless further analysis reveals they were actually supposed to be unique records. We'll pick the 1st record in the pair to keep.

In [0]:
voterInfoDF[voterInfoDF.duplicated(keep = False)]  # Display a portion of the duplicated records, if there are any.

SOS_VOTERID  Individual.Id  Gender  ...  Arab Jewish  Mediterranean
465    OH0018759880      515495171    Male  ...     0      0              0
466    OH0018759880      515495171    Male  ...     0      0              0
976             NaN      515508941  Female  ...     0      0              0
977             NaN      515508941  Female  ...     0      0              0
1522   OH0019065624      515493026  Female  ...     0      0              0
...             ...            ...     ...  ...   ...    ...            ...
18663  OH0022833156      826670128     NaN  ...     0      0              0
19260  OH0012701841      104028131  Female  ...     0      0              0
19261  OH0012701841      104028131  Female  ...     0      0              0
19317           NaN      784215921  Female  ...     0      0              0
19318           NaN      784215921  Female  ...     0      0              0

[270 rows x 18 columns]

In [0]:
# Uncomment this line to save the duplicates to a file.
#voterInfoDF[voterInfoDF.duplicated(keep = False)].to_csv("duplicate_records.csv")
# Switch keep to first so we keep one of each pair.
voterInfoDF.drop(index = voterInfoDF[voterInfoDF.duplicated(keep = 'first')].index, inplace = True)
voterInfoDF.shape

(20340, 16)

**At this point, we'll split into labeled and unlabeled dataframes. Only labeled will be used for traning**  


In [0]:
labeledDF = voterInfoDF[voterInfoDF["PARTY_AFFILIATION"].notna()].copy()
not_labeledDF = voterInfoDF[voterInfoDF["PARTY_AFFILIATION"].isna()].copy()

In [0]:
print("INFO: Un-Labeled data records:", not_labeledDF.index.size)
print("INFO: Labeled data records:", labeledDF.index.size)
# unlabeled records may exist in train mode. They will be the entire file in predict mode.
# Note that this assumes unlabled records will be NaN in the _REGCOL (Registered or LALVOTERID)
labeledDF["PARTY_AFFILIATION"].value_counts()

INFO: Un-Labeled data records: 15726
INFO: Labeled data records: 4614


D    2427
R    2187
Name: PARTY_AFFILIATION, dtype: int64

**Convert Party and Gender to numerical values for ML models to work with more easily.**

In [0]:
mappingDR = {'D':1, 'R':0, 'Male':0, 'Female':1}
labeledDF.replace(to_replace=mappingDR, inplace = True)
not_labeledDF.replace(to_replace=mappingDR, inplace = True)

In [0]:
TotalDemsInData = labeledDF.query("PARTY_AFFILIATION == 1").index.size
TotalNotDemsInData = labeledDF.query("PARTY_AFFILIATION == 0").index.size
print("INFO: Number of Dem records in labeled data:", TotalDemsInData)
print("INFO: Percent of Dems in labeled population: {:.1%}".format(TotalDemsInData/labeledDF.index.size))

INFO: Number of Dem records in labeled data: 2427
INFO: Percent of Dems in labeled population: 52.6%


**We've got some missing values we'll need to impute when prepping for ML**  
`.info()` shows the number of non-null values for each column.
This will be easier to do in the modeling state since SciKit supports multiple methods and we can try different methods to see what works best.

In [0]:
labeledDF.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 4614 entries, 5 to 20484
Data columns (total 16 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Gender                    4564 non-null   float64
 1   Inferred.Age              4353 non-null   float64
 2   PARTY_AFFILIATION         4614 non-null   int64  
 3   percentBlack              4613 non-null   float64
 4   firstNameNumVoters        4605 non-null   float64
 5   firstNamefractionDem      4605 non-null   float64
 6   firstNameOccurenceBinned  4614 non-null   int64  
 7   lastNameNumVoters         4583 non-null   float64
 8   lastNamefractionDem       4583 non-null   float64
 9   apartmentDweller          4614 non-null   int64  
 10  AA                        4614 non-null   int64  
 11  Hispanic                  4614 non-null   int64  
 12  Asian                     4614 non-null   int64  
 13  Arab                      4614 non-null   int64  
 14  Jewish 

**Now check correlation to the Party Affiliation for with other columns (labeled data only)**  

In [0]:
print("INFO: Pearson correlation coefficients. SR values should show high positive correlation")
# Uncomment the next line to save an updated correlation matrix
#pd.DataFrame(labeledDF.corr('pearson')).to_csv("fullCorrMatrix2.csv")
print(labeledDF.corr('pearson')) # Correlation matrix. Note that we're only looking at the labeled data here.

INFO: Pearson correlation coefficients. SR values should show high positive correlation
                            Gender  Inferred.Age  ...    Jewish  Mediterranean
Gender                    1.000000      0.033460  ... -0.001299      -0.010688
Inferred.Age              0.033460      1.000000  ...  0.005457       0.010133
PARTY_AFFILIATION         0.102354     -0.039412  ... -0.052725      -0.025421
percentBlack              0.042000     -0.012232  ... -0.026009      -0.048200
firstNameNumVoters       -0.346798      0.122335  ... -0.007558      -0.013657
firstNamefractionDem      0.314310     -0.124199  ... -0.038839      -0.028864
firstNameOccurenceBinned -0.071877      0.089452  ...  0.001655       0.025349
lastNameNumVoters         0.008117     -0.044192  ... -0.063624      -0.062295
lastNamefractionDem       0.047677     -0.027076  ... -0.057206       0.015845
apartmentDweller         -0.001755      0.028311  ... -0.004550      -0.007381
AA                        0.041495     -0.0

**Looking at a scatter plot matrix of those variables is helpful to.**

In [0]:
from pandas.plotting import scatter_matrix
columnsForScatterPlot = ['Inferred.Age','PARTY_AFFILIATION', 'percentBlack', 'firstNameNumVoters',
                         'firstNamefractionDem','firstNameOccurenceBinned','lastNameNumVoters', 
                         'lastNamefractionDem','AA']
plotDF = labeledDF[columnsForScatterPlot]
# Commented, takes a bit to run.
#scatter_matrix(plotDF, alpha=0.2, figsize = (12,9))

## Train and Validate the Model

**Start with building a pipeline to impute values and normalize values, then model and validate**

In [0]:
from sklearn.impute import SimpleImputer
#from sklearn.impute import IterativeImputer
from sklearn.preprocessing import MinMaxScaler
from sklearn.pipeline import Pipeline
from sklearn.compose import ColumnTransformer
from sklearn.ensemble import RandomForestClassifier
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.svm import LinearSVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn.model_selection import train_test_split
from sklearn.model_selection import GridSearchCV
from sklearn.model_selection import cross_val_score
from sklearn.metrics import accuracy_score

**Impute**  
Can use a strategy like mean or median on the float columns, but anything with binary or categorical data we'll want to use the constant strategy and pick a value that will be neutral or lean democrat. The KNNImputer or IterativeImputer might be a good option to try in place of the SimpleImputer.

A problem we might run into with SimpleImputer is that many of these variables are related to each other. If our mean `firstNamefractionDem` value doesn't make sense relative to the `firstNameNumVoters` value for that same record, that could throw off the model. We would essentially be injecting an outlier. So maybe KNN or Iterative can help with that, too.

Continuous:  
* Inferred.Age              4353 non-null   float64  
* percentBlack              4613 non-null   float64  
* firstNameNumVoters        4605 non-null   float64  
* firstNamefractionDem      4605 non-null   float64  
* lastNameNumVoters         4583 non-null   float64  
* lastNamefractionDem       4583 non-null   float64  

Discrete:  
* Gender                    4564 non-null: Could try randomly 0 or 1, or use IterativeImputer to predict the gender based on the name. Might try that later.

In [0]:
#These are't great, but they'll get things working. It might even be better to drop missing values then impute iffy values
imputerContinuous = SimpleImputer(strategy="median")
imputerDiscrete = SimpleImputer(strategy="constant", fill_value=0)

#imputerDiscrete = IterativeImputer(random_state=0, n_nearest_features=2, initial_strategy='constant', 
#                                   verbose=2)
#fill_value=0
contColsImpute = ['Inferred.Age', 'percentBlack', 'firstNameNumVoters', 'firstNamefractionDem',
                    'lastNameNumVoters','lastNamefractionDem']
discColsImpute = ['Gender'] # Keep these separate from the other discrete binary cols since imput must be handled case by case
discOrdCols = ['firstNameOccurenceBinned']
otherCols = ['apartmentDweller','AA','Hispanic','Asian','Arab','Jewish','Mediterranean']

# Quick error check. We're going to use these cols for input, so to make sure we didn't forget one, 
# compare to the labeled DF. Don't forget to account for the target columnn
if (len(contColsImpute)+
    len(discColsImpute)+
    len(discOrdCols)+
    len(otherCols)) != (len(list(labeledDF.columns)) - 1):
    print("WARNING: Number of cols going into model doesn't match labeled DataFrame columns")

In [0]:
imputerContinuous.fit(labeledDF[contColsImpute])
imputerContinuous.statistics_

array([6.100e+01, 1.590e-01, 6.058e+03, 3.937e-01, 4.140e+02, 4.239e-01])

In [0]:
imputerDiscrete.fit(labeledDF[discColsImpute])
imputerDiscrete.statistics_

array([0.])

**Normalize**  
Scale everything to the same 0..1 range for better model behavior. Don't really want to standardize (match the means) because I'm not sure how models will deal with the binary values in the discrete columns

In [0]:
normalizer = MinMaxScaler()
normalizer.fit(labeledDF[contColsImpute])

MinMaxScaler(copy=True, feature_range=(0, 1))

**Pipeline**  
We need to be able to do all of this imputing and normalizing when we get to the unlabeled data as well, so package this process up into a pipeline that can be used with a model.

In [0]:
pipeContinuous = Pipeline([('impute', SimpleImputer(strategy="median")),
                           ('normalize', MinMaxScaler())])
pipeDiscrete = Pipeline([('imputeDisc', SimpleImputer(strategy="constant", fill_value=0))])

fullPipe = ColumnTransformer([
    ("cont", pipeContinuous, contColsImpute),
    ("disc", pipeDiscrete, discColsImpute)],
    remainder = 'passthrough')

inputDF = labeledDF.drop(columns=['PARTY_AFFILIATION'])
targetDF = labeledDF['PARTY_AFFILIATION']

dataModelReady = fullPipe.fit_transform(inputDF)

In [0]:
remainderCols = [col for col in list(inputDF) if col not in contColsImpute + discColsImpute]

In [0]:
pd.DataFrame(dataModelReady, columns = contColsImpute + discColsImpute + remainderCols)

Inferred.Age  percentBlack  ...  Jewish  Mediterranean
0         0.282051      0.091825  ...     0.0            0.0
1         0.064103      0.041433  ...     0.0            0.0
2         0.153846      0.910414  ...     0.0            0.0
3         0.115385      0.095185  ...     0.0            0.0
4         0.256410      0.090705  ...     0.0            0.0
...            ...           ...  ...     ...            ...
4609      0.282051      0.178052  ...     0.0            0.0
4610      0.115385      0.415454  ...     0.0            0.0
4611      0.256410      0.108623  ...     0.0            0.0
4612      0.294872      0.010078  ...     0.0            1.0
4613      0.666667      0.363942  ...     0.0            0.0

[4614 rows x 15 columns]

## Model Training and Scoring

**KNN**

In [0]:
knn = KNeighborsClassifier()
fullPipe_KNN = Pipeline([
                        ('dataPrep', fullPipe),
                        ('model', knn)])

scores = cross_val_score(fullPipe_KNN, inputDF, targetDF, scoring = 'accuracy', cv = 5)
print("KNN classifier score:", scores)

KNN classifier score: [0.61646804 0.59371614 0.57746479 0.57638137 0.63123644]


**Support Vector Classifier**

In [0]:
svc = LinearSVC(random_state=0, tol=1e-5)
fullPipe_SVC = Pipeline([
                        ('dataPrep', fullPipe),
                        ('model', svc)])

scores = cross_val_score(fullPipe_SVC, inputDF, targetDF, scoring = 'accuracy', cv = 5)
print("SVC classifier score:", scores)

SVC classifier score: [0.67063922 0.64788732 0.67063922 0.65655471 0.70173536]


**Random Forest**

In [0]:
rfc = RandomForestClassifier(max_depth=3, random_state=0)
fullPipe_RFC = Pipeline([
                        ('dataPrep', fullPipe),
                        ('model', rfc)])

scores = cross_val_score(fullPipe_RFC, inputDF, targetDF, scoring = 'accuracy', cv = 5)
print("RFC classifier score:", scores)

RFC classifier score: [0.66305525 0.66197183 0.66630553 0.64355363 0.69197397]


**Gradient Boosting**

In [0]:
params = {'n_estimators': 400, 'max_depth': 3, 'subsample': 0.5,
          'learning_rate': 0.01, 'min_samples_leaf': 1, 'random_state': 3}
gbc = GradientBoostingClassifier(**params)
fullPipe_GBC = Pipeline([
                        ('dataPrep', fullPipe),
                        ('model', gbc)])

scores = cross_val_score(fullPipe_GBC, inputDF, targetDF, scoring = 'accuracy', cv = 5)
print("GBC classifier score:", scores)

GBC classifier score: [0.67822319 0.65872156 0.66305525 0.65222102 0.70173536]


## Make Predictions, save result file

**First, separate our unlabeled data to be predicted into input and output columns like we did with the labeled data**

In [0]:
unlabeled_inputDF = not_labeledDF.drop(columns=['PARTY_AFFILIATION'])
unlabeled_inputDF.describe()

Gender  Inferred.Age  ...        Jewish  Mediterranean
count  15370.000000   14123.00000  ...  15726.000000   15726.000000
mean       0.529668      48.64002  ...      0.022383       0.011828
std        0.499135      18.17423  ...      0.147931       0.108113
min        0.000000      20.00000  ...      0.000000       0.000000
25%        0.000000      33.00000  ...      0.000000       0.000000
50%        1.000000      46.00000  ...      0.000000       0.000000
75%        1.000000      62.00000  ...      0.000000       0.000000
max        1.000000      99.00000  ...      1.000000       1.000000

[8 rows x 15 columns]

In [0]:
fullPipe_SVC.fit(inputDF, targetDF) # This will train on the full dataset, not just the n-folds during from validation
                           # Both of those dataframes are the labeled data

Pipeline(memory=None,
         steps=[('dataPrep',
                 ColumnTransformer(n_jobs=None, remainder='passthrough',
                                   sparse_threshold=0.3,
                                   transformer_weights=None,
                                   transformers=[('cont',
                                                  Pipeline(memory=None,
                                                           steps=[('impute',
                                                                   SimpleImputer(add_indicator=False,
                                                                                 copy=True,
                                                                                 fill_value=None,
                                                                                 missing_values=nan,
                                                                                 strategy='median',
                                                          

In [0]:
predictedParty = fullPipe_SVC.predict(unlabeled_inputDF)

In [0]:
resultDF = pd.DataFrame(predictedParty, columns = ['PredictedParty']).join(unlabeled_inputDF)
resultDF.head()

PredictedParty  Gender  Inferred.Age  ...  Arab  Jewish  Mediterranean
0               0     NaN           NaN  ...   NaN     NaN            NaN
1               0     0.0          49.0  ...   0.0     0.0            0.0
2               0     0.0          64.0  ...   0.0     0.0            0.0
3               0     NaN          87.0  ...   0.0     0.0            0.0
4               0     NaN          87.0  ...   0.0     0.0            0.0

[5 rows x 16 columns]

In [0]:
'''TotalDemsInData = not_labeledDF["dem"].value_counts().loc[1]
TotalNotDemsInData = not_labeledDF["dem"].value_counts().loc[0]
print("INFO: Number of Dem records predicted in unlabeled data:", TotalDemsInData)
print("INFO: Percent of Dems predicted in unlabeled population: {:.1%}".format(TotalDemsInData/not_labeledDF.index.size))'''

'TotalDemsInData = not_labeledDF["dem"].value_counts().loc[1]\nTotalNotDemsInData = not_labeledDF["dem"].value_counts().loc[0]\nprint("INFO: Number of Dem records predicted in unlabeled data:", TotalDemsInData)\nprint("INFO: Percent of Dems predicted in unlabeled population: {:.1%}".format(TotalDemsInData/not_labeledDF.index.size))'

In [0]:
filenameTS = "./NBVP_data_out/dem_prediction_data_"+"{:%Y-%m-%d_%H%M%S}".format(datetime.datetime.now())+".csv"
try:
    resultDF.to_csv(filenameTS, index=False) # Don't need the Index values here.
except:
    print("INFO: Writing prediction data to ", filenameTS)